### PR TITLE
WIP: Protoyping. Investigate how to monitor core components with statsd

### PR DESCRIPTION
**Pull request by fxulusoy**
Prototype 1.
Refactoring.
Fix.
Remove whitespaces.

Signed-off-by: Christian Bickel <cbickel@de.ibm.com>

### CODE REVIEW

**Comment by fxulusoy** [comment]
@rabbah @ioana-blue We can discuss based on this PR.

File: common/scala/src/whisk/common/Logging.scala [hunk 1]
```diff
@@ -42,25 +44,25 @@ trait Logging {
     def setComponentName(comp: String) =
         this.componentName = comp
 
-    def debug(from: AnyRef, message: String, marker: String = null)(implicit id: TransactionId = TransactionId.unknown) = {
+    def debug(from: AnyRef, message: String, marker: LogMarkerToken = null)(implicit id: TransactionId = TransactionId.unknown) = {
         val mark = id.mark(marker)
```
Comment (ioana-blue) [inline common/scala/src/whisk/common/Logging.scala]: Is it possible to add the markers you intend to use to the predefined list so you don't need to change all these to String (and continue to use LogMarkerToken instead)?

**Comment by rabbah** [comment]
Why are we doing this in the core? Wasn't the plan to do push from the logstash forwarder using the statsd plugin?

**Comment by fxulusoy** [comment]
@rabbah we agree that we can move statsd call into the Logstash. we also thought that statsd calls in transactionId is not the best way. Our idea is to calculate deltas in the transaction context and just do send action in the Logstash.

**Comment by ioana-blue** [comment]
You may be able to do the deltas in Logstash as well, so you could move all the logic to Logstash. I think for the use cases you are interested in, the log messages appear in the same file. If that's the case, you can use this plugin: [elapsed](https://www.elastic.co/guide/en/logstash/current/plugins-filters-elapsed.html) and compute the deltas and then send them to statsd.

**Comment by fxulusoy** [comment]
@ioana-blue @rabbah We want to do more fine-grained metrics. Not a duration of whole transaction or from transaction start to some point in a transaction. We also want to measure duration of external calls separately. After the first implementation from Ioana, we did not consider doing calculation in the Logstash. Because we thought that it is not possible to do that in the Logstash. Btw, we are not experienced with Logstash. Thus we implemented in the transaction context.

**Comment by ioana-blue** [comment]
@fxulusoy Give Logstash a try. It's fairly simple to use the elapsed plugin. I didn't know Logstash either, but you can read some of the documentation and it's fairly straightforward. Take a look at the plugin and let me know if you have questions, we can chat over Skype if necessary.

**Comment by perryibm** [comment]
@fxulusoy The reason we want to offload from the controller is that it is at the central component and performance critical.  By emitting one UDP packet  per log, we would end up generating much more traffic in service os a single activation than the one call coming in.  So offloading work form the controller and some sort of batching is important and why I think we are pushing Logstash.

**Comment by fxulusoy** [comment]
@perryibm I totally agree with your concerns about performance. As a said in one of my previous comment, making UDP calls per log is wrong. The PR is just a prototype to show our idea and we just put UPD calls there for a moment. The goal of PR is to show how we want to calculate the deltas not how /where to call statsd. we will have a look logstash and the related plugin and give a try.

**Pull request closed without merging**